File: scripts/debug_stripchat_api.py
import json
import os
import subprocess
import sys
from urllib.parse import urlparse

try:
    from playwright.sync_api import sync_playwright
    HAS_PY_PLAYWRIGHT = True
except Exception:
    sync_playwright = None
    HAS_PY_PLAYWRIGHT = False


UA = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36"
)


def _short_headers(headers):
    keep = (
        "user-agent",
        "referer",
        "origin",
        "accept",
        "cookie",
        "sec-fetch-site",
        "sec-fetch-mode",
    )
    out = {}
    for key, value in headers.items():
        lower = key.lower()
        if lower in keep:
            if lower == "cookie" and len(value) > 200:
                out[key] = value[:200] + "...<trimmed>"
            else:
                out[key] = value
    return out


def run(url):
    if not HAS_PY_PLAYWRIGHT:
        script_dir = os.path.dirname(os.path.abspath(__file__))
        js_path = os.path.join(script_dir, "debug_stripchat_api.js")
        print("[*] Python Playwright not installed, falling back to Node: {}".format(js_path))
        result = subprocess.run(
            ["node", js_path, url],
            cwd=script_dir,
            check=False,
        )
        if result.returncode != 0:
            raise SystemExit(result.returncode)
        return

    print("[*] Analyzing browser traffic for: {}".format(url))
    print("[*] This runs outside Kodi and prints the real browser media requests.")
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(
            viewport={"width": 1440, "height": 900},
            user_agent=UA,
        )
        page = context.new_page()

        seen = set()
        media_events = []
        target_host = urlparse(url).netloc.lower()
        stop_early = {"armed": False}

        def log_media_event(kind, request, response=None, body_text=None):
            req_url = request.url
            if req_url in seen:
                return
            seen.add(req_url)

            lower = req_url.lower()
            if not any(ext in lower for ext in (".m3u8", ".mp4", ".m4s")):
                return
            if any(x in lower for x in ("/thumbs/", "/images/", ".jpg", ".png", "/image/", "/thumb/")):
                return

            event = {
                "kind": kind,
                "url": req_url,
                "method": request.method,
                "resource_type": request.resource_type,
                "headers": _short_headers(request.headers),
            }
            if response is not None:
                event["status"] = response.status
                event["response_headers"] = _short_headers(response.headers)
            if body_text is not None:
                event["body_preview"] = body_text[:800]

            media_events.append(event)
            if response is not None and response.status == 200:
                stop_early["armed"] = True

            print("\n[MEDIA {}] {}".format(kind.upper(), req_url))
            print("  method={} type={}".format(request.method, request.resource_type))
            if response is not None:
                print("  status={}".format(response.status))
            if event["headers"]:
                print("  request_headers={}".format(json.dumps(event["headers"], indent=2)))
            if response is not None and event.get("response_headers"):
                print(
                    "  response_headers={}".format(
                        json.dumps(event["response_headers"], indent=2)
                    )
                )
            if body_text:
                if "#EXT-X-MOUFLON" in body_text:
                    print("  manifest_type=MOUFLON")
                elif "#EXTM3U" in body_text:
                    print("  manifest_type=STANDARD")
                print("  body_preview={}".format(body_text[:400].replace("\n", "\\n")))

        def handle_request(request):
            lower = request.url.lower()
            if target_host in lower and "widget" in lower:
                print("\n[API REQUEST] {}".format(request.url))
            if any(ext in lower for ext in (".m3u8", ".mp4", ".m4s")):
                log_media_event("request", request)

        def handle_response(response):
            request = response.request
            try:
                lower = response.url.lower()
                ct = response.headers.get("content-type", "").lower()

                if target_host in lower and "application/json" in ct:
                    try:
                        data = response.json()
                        if "stream" in str(data).lower() or "hls" in str(data).lower():
                            print("\n[API JSON] {}".format(response.url))
                            print(json.dumps(data, indent=2)[:1500])
                    except Exception:
                        pass

                body_text = None
                if ".m3u8" in lower:
                    try:
                        body_text = response.text()
                    except Exception:
                        body_text = None

                if any(ext in lower for ext in (".m3u8", ".mp4", ".m4s")):
                    log_media_event("response", request, response=response, body_text=body_text)
            except Exception as exc:
                print("[!] response handler error: {}".format(exc))

        page.on("request", handle_request)
        page.on("response", handle_response)

        try:
            page.goto(url, wait_until="domcontentloaded", timeout=30000)
            print("[*] Page loaded")
            page.wait_for_timeout(3000)

            play_selectors = [
                "video",
                "button",
                ".vjs-big-play-button",
                ".play-button",
                "[data-test='video-play-button']",
            ]
            for selector in play_selectors:
                try:
                    locator = page.locator(selector).first
                    if locator.is_visible(timeout=1000):
                        print("[*] Clicking {}...".format(selector))
                        locator.click(force=True, timeout=3000)
                        page.wait_for_timeout(2000)
                except Exception:
                    continue

            print("[*] Waiting up to 15s for media traffic...")
            for _ in range(15):
                page.wait_for_timeout(1000)
                if stop_early["armed"]:
                    print("[*] Captured successful media response, stopping early.")
                    break

            print("\n[*] Media event count: {}".format(len(media_events)))
            print("[*] Unique media URLs captured:")
            for event in media_events:
                print(" - {} {} {}".format(
                    event["kind"].upper(),
                    event.get("status", ""),
                    event["url"],
                ).rstrip())
        finally:
            browser.close()


if __name__ == "__main__":
    target = sys.argv[1] if len(sys.argv) > 1 else "https://stripchat.com/RussianWoman"
    run(target)

File: scripts/debug_stripchat_api.js
const { chromium } = require('@playwright/test');

const UA =
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 ' +
  '(KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36';

function shortHeaders(headers) {
  const keep = new Set([
    'user-agent',
    'referer',
    'origin',
    'accept',
    'cookie',
    'sec-fetch-site',
    'sec-fetch-mode',
  ]);
  const out = {};
  for (const [key, value] of Object.entries(headers || {})) {
    const lower = key.toLowerCase();
    if (!keep.has(lower)) continue;
    if (lower === 'cookie' && typeof value === 'string' && value.length > 200) {
      out[key] = value.slice(0, 200) + '...<trimmed>';
    } else {
      out[key] = value;
    }
  }
  return out;
}

function shouldCapture(url) {
  const lower = String(url || '').toLowerCase();
  if (!['.m3u8', '.mp4', '.m4s'].some((ext) => lower.includes(ext))) return false;
  if (['/thumbs/', '/images/', '.jpg', '.png', '/image/', '/thumb/'].some((x) => lower.includes(x))) return false;
  return true;
}

async function run(url) {
  console.log(`[*] Analyzing browser traffic for: ${url}`);
  console.log('[*] This runs outside Kodi and prints the real browser media requests.');

  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    viewport: { width: 1440, height: 900 },
    userAgent: UA,
  });
  const page = await context.newPage();

  const seen = new Set();
  const mediaEvents = [];
  const targetHost = new URL(url).host.toLowerCase();
  let stopEarly = false;

  function logMediaEvent(kind, request, response = null, bodyText = null) {
    const reqUrl = request.url();
    if (seen.has(`${kind}:${reqUrl}`)) return;
    if (!shouldCapture(reqUrl)) return;
    seen.add(`${kind}:${reqUrl}`);

    const event = {
      kind,
      url: reqUrl,
      method: request.method(),
      resource_type: request.resourceType(),
      headers: shortHeaders(request.headers()),
    };
    if (response) {
      event.status = response.status();
      event.response_headers = shortHeaders(response.headers());
    }
    if (bodyText) {
      event.body_preview = bodyText.slice(0, 800);
    }
    mediaEvents.push(event);
    if (response && response.status() === 200) {
      stopEarly = true;
    }

    console.log(`\n[MEDIA ${kind.toUpperCase()}] ${reqUrl}`);
    console.log(`  method=${event.method} type=${event.resource_type}`);
    if (typeof event.status !== 'undefined') {
      console.log(`  status=${event.status}`);
    }
    if (Object.keys(event.headers).length) {
      console.log(`  request_headers=${JSON.stringify(event.headers, null, 2)}`);
    }
    if (event.response_headers && Object.keys(event.response_headers).length) {
      console.log(`  response_headers=${JSON.stringify(event.response_headers, null, 2)}`);
    }
    if (bodyText) {
      if (bodyText.includes('#EXT-X-MOUFLON')) {
        console.log('  manifest_type=MOUFLON');
      } else if (bodyText.includes('#EXTM3U')) {
        console.log('  manifest_type=STANDARD');
      }
      console.log(`  body_preview=${bodyText.slice(0, 400).replace(/\n/g, '\\n')}`);
    }
  }

  page.on('request', (request) => {
    const lower = request.url().toLowerCase();
    if (lower.includes(targetHost) && lower.includes('widget')) {
      console.log(`\n[API REQUEST] ${request.url()}`);
    }
    if (shouldCapture(request.url())) {
      logMediaEvent('request', request);
    }
  });

  page.on('response', async (response) => {
    const request = response.request();
    try {
      const lower = response.url().toLowerCase();
      const headers = response.headers();
      const ct = String(headers['content-type'] || '').toLowerCase();

      if (lower.includes(targetHost) && ct.includes('application/json')) {
        try {
          const data = await response.json();
          const asText = JSON.stringify(data);
          if (asText.toLowerCase().includes('stream') || asText.toLowerCase().includes('hls')) {
            console.log(`\n[API JSON] ${response.url()}`);
            console.log(JSON.stringify(data, null, 2).slice(0, 1500));
          }
        } catch (_) {}
      }

      let bodyText = null;
      if (lower.includes('.m3u8')) {
        try {
          bodyText = await response.text();
        } catch (_) {}
      }

      if (shouldCapture(response.url())) {
        logMediaEvent('response', request, response, bodyText);
      }
    } catch (err) {
      console.log(`[!] response handler error: ${err}`);
    }
  });

  try {
    await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 30000 });
    console.log('[*] Page loaded');
    await page.waitForTimeout(3000);

    const playSelectors = [
      'video',
      'button',
      '.vjs-big-play-button',
      '.play-button',
      "[data-test='video-play-button']",
    ];
    for (const selector of playSelectors) {
      try {
        const locator = page.locator(selector).first();
        if (await locator.isVisible({ timeout: 1000 })) {
          console.log(`[*] Clicking ${selector}...`);
          await locator.click({ force: true, timeout: 3000 });
          await page.waitForTimeout(2000);
        }
      } catch (_) {}
    }

    console.log('[*] Waiting up to 15s for media traffic...');
    for (let i = 0; i < 15; i++) {
      await page.waitForTimeout(1000);
      if (stopEarly) {
        console.log('[*] Captured successful media response, stopping early.');
        break;
      }
    }

    console.log(`\n[*] Media event count: ${mediaEvents.length}`);
    console.log('[*] Unique media URLs captured:');
    for (const event of mediaEvents) {
      const status = typeof event.status === 'undefined' ? '' : `${event.status}`;
      console.log(` - ${event.kind.toUpperCase()} ${status} ${event.url}`.trim());
    }
  } finally {
    await browser.close();
  }
}

const target = process.argv[2] || 'https://stripchat.com/RussianWoman';
run(target).catch((err) => {
  console.error(err);
  process.exit(1);
});
